File: java/lib/src/test/java/io/v/v23/security/CaveatTest.java
// Copyright 2015 The Vanadium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package io.v.v23.security;

import static com.google.common.truth.Truth.assertThat;

import junit.framework.TestCase;

import org.joda.time.DateTime;

import io.v.v23.V;
import io.v.v23.context.VContext;
import io.v.v23.verror.VException;
import io.v.x.jni.test.security.TestCaveatValidator;

import java.util.Arrays;

/**
 * Tests the various caveat implementations.
 */
public class CaveatTest extends TestCase {
    public void testMethodCaveat() throws VException {
        VContext context = V.init();
        VPrincipal p1 = VSecurity.newPrincipal();
        Blessings alice = p1.blessSelf("alice", VSecurity.newMethodCaveat("Succeed"));
        VSecurity.addToRoots(p1, alice);
        {
            Call call = VSecurity.newCall(
                    new CallParams().withLocalPrincipal(p1).withRemoteBlessings(alice).withMethod("succeed"));
            String[] result = VSecurity.getRemoteBlessingNames(context, call);
            assertThat(Arrays.asList(result)).containsExactly("alice");
        }
        {
            Call call = VSecurity.newCall(
                    new CallParams().withLocalPrincipal(p1).withMethod("Fail"));
            String[] result = VSecurity.getRemoteBlessingNames(context, call);
            assertThat(result != null).isTrue();
            assertThat(Arrays.asList(result)).containsExactly();
        }
    }

    public void testExpiryCaveat() throws VException {
        VContext context = V.init();
        VPrincipal p1 = VSecurity.newPrincipal();
        Blessings alice = p1.blessSelf(
            "alice", VSecurity.newExpiryCaveat(DateTime.now().plusHours(1)));
        VSecurity.addToRoots(p1, alice);
        {
            Call call = VSecurity.newCall(new CallParams()
                    .withLocalPrincipal(p1)
                    .withRemoteBlessings(alice)
                    .withTimestamp(DateTime.now()));

            String[] result = VSecurity.getRemoteBlessingNames(context, call);
            assertThat(Arrays.asList(result)).containsExactly("alice");
        }
        {
            Call call = VSecurity.newCall(new CallParams()
                    .withLocalPrincipal(p1)
                    .withTimestamp(DateTime.now().plusHours(2)));
            String[] result = VSecurity.getRemoteBlessingNames(context, call);
            assertThat(result != null).isTrue();
            assertThat(Arrays.asList(result)).containsExactly();
        }
    }

    public void testCustomCaveat() throws VException {
        VContext context = V.init();
        CaveatRegistry.register(io.v.x.jni.test.security.Constants.TEST_CAVEAT,
                new TestCaveatValidator());
        VPrincipal p1 = VSecurity.newPrincipal();
        Blessings alice = p1.blessSelf("alice",
                VSecurity.newCaveat(io.v.x.jni.test.security.Constants.TEST_CAVEAT, "succeed"));
        VSecurity.addToRoots(p1, alice);
        {
            Call call = VSecurity.newCall(new CallParams()
                    .withLocalPrincipal(p1)
                    .withRemoteBlessings(alice)
                    .withSuffix("succeed"));
            String[] result = VSecurity.getRemoteBlessingNames(context, call);
            assertThat(Arrays.asList(result)).containsExactly("alice");
        }
        {
            Call call = VSecurity.newCall(new CallParams()
                    .withLocalPrincipal(p1)
                    .withRemoteBlessings(alice)
                    .withSuffix("fail"));
            String[] result = VSecurity.getRemoteBlessingNames(context, call);
            assertThat(result != null).isTrue();
            assertThat(Arrays.asList(result)).containsExactly();
        }
    }
}

File: go/src/v.io/v23/security/access/caveat_test.go
// Copyright 2015 The Vanadium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package access_test

import (
	"testing"

	"v.io/v23/context"
	"v.io/v23/security"
	"v.io/v23/security/access"
	"v.io/v23/vdl"
)

func TestAccessTagCaveat(t *testing.T) {
	var (
		server     = newPrincipal(t)
		bserver, _ = server.BlessSelf("server")
		caveat, _  = access.NewAccessTagCaveat(access.Debug, access.Resolve)
		bclient, _ = server.Bless(newPrincipal(t).PublicKey(), bserver, "debugger", caveat)
		tests      = []struct {
			MethodTags []*vdl.Value
			OK         bool
		}{
			{nil, false},
			{[]*vdl.Value{vdl.ValueOf(access.Debug)}, true},
			{[]*vdl.Value{vdl.ValueOf(access.Resolve)}, true},
			{[]*vdl.Value{vdl.ValueOf(access.Read), vdl.ValueOf(access.Debug)}, true},
			{[]*vdl.Value{vdl.ValueOf(access.Read), vdl.ValueOf(access.Write)}, false},
			{[]*vdl.Value{vdl.ValueOf("Debug"), vdl.ValueOf("Resolve")}, false},
		}
	)
	security.AddToRoots(server, bserver)
	ctx, cancel := context.RootContext()
	defer cancel()
	for idx, test := range tests {
		call := security.NewCall(&security.CallParams{
			MethodTags:      test.MethodTags,
			LocalPrincipal:  server,
			RemoteBlessings: bclient,
		})
		got, rejected := security.RemoteBlessingNames(ctx, call)
		if test.OK {
			if len(got) != 1 || got[0] != "server:debugger" {
				t.Errorf("Got (%v, %v), wanted ([%q], nil) for method tags %v (test case #%d)", got, rejected, "server:debugger", test.MethodTags, idx)
			}
		}
		if !test.OK && len(got) != 0 {
			t.Errorf("Got (%v, %v), wanted all blessings to be rejected for method tags %v (test case #%d)", got, rejected, test.MethodTags, idx)
		}
	}
}
